### ISSUE TITLE
Does TF_SessionRun have a memory leak?

### ISSUE THREAD

**Issue by el1995** [tensorflow/tensorflow#29733]
<em>Please make sure that this is a bug. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes, it is a C++ project using the C API
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 16.04 LTS
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: no
- TensorFlow installed from (source or binary): C API (no installation necessary, only linked)
- TensorFlow version (use command below): 1.13
- Python version: 3.6
- Bazel version (if compiling from source): -
- GCC/Compiler version (if compiling from source): -
- CUDA/cuDNN version: -
- GPU model and memory: - (CPU only)

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with: 1. TF 1.0: `python -c "import
tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"` 2. TF 2.0: `python -c
"import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**

I have written a code performing inference on frozen graphs. Due to certain circumstances I must use the C API. The code does exactly what I expect. However as I must do the inference thousands of times during a simulation, at some point the simulation crashes as I run out of memory. So I started to uncomment my code and rejoin one part after the other. I found out that the memory leak only occurs if I uncomment TF_SessionRun. I was unable to figure out what it does to increase the required memory. Maybe it is also my fault but I do not see an error in memory management. Another remarkable thing is that before the inference I get the following warning 5 times:

`W tensorflow/core/framework/allocator.cc:124] Allocation of 7420800000 exceeds 10% of system memory.`

**Describe the expected behavior**

Basically just as it is now, without increasing used memory every single time I do the inference.

**Code to reproduce the issue**
I will come up later with the code.

**Other info / logs**
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.

**Comment by achandraa** [comment]
Please provide us the minimal code snippet that can replicate the issue. This will help us to proceed faster. Thanks!

**Comment by VladislavAD** [comment]
Hi, I have similar issue with `TF_SessionRun` on Windows 10 with binaries Tensorflow 1.14, I'm also using Visual Studio Community 2017 and its heap profiling tool. I'm doing semantic segmentation using U-Net and every time I run `TF_SessionRun` I have leak of size 
(input tensor dimension size) + (output tensor dimension size) + 159 bytes. 
So for example when I predict one class 512x512 RGB image I have 512 * 512 * 3 + 512 * 512 * 1 + 159 = 1048735 bytes leaked. All this leaked memory remain after `TF_DeleteSession` (but I don't want to reopen session every time I want to predict, first prediction takes too much time). The code snippet:

```c
	float * arrayOfFloats = (float*)malloc(512 * 512 * 3 * sizeof(float));
	for (int i = 0; i < 512 * 512 * 3; i++) {
		arrayOfFloats[i] = 100;
	}
	const int64_t dims[4] = { 1, 512, 512, 3 };
	size_t nbytes = 512 * 512 * 3;

	TF_Output* input;
	TF_Tensor* input_value;
	TF_Output input_opout = { model->input.oper, 0 };
	input = &input_opout;
	input_value = TF_NewTensor(TF_FLOAT, dims, 4, (void*)arrayOfFloats, nbytes * sizeof(float), &Deallocator, NULL);

	int outputSize = 512 * 512;

	int64_t out_dims[] = { 1, 512, 512, 1 };
	TF_Output* output;
	TF_Output output_opout = { model->output.oper, 0 };
	output = &output_opout;
	TF_Tensor* output_value = TF_AllocateTensor(TF_FLOAT, out_dims, 4, outputSize * sizeof(float));

	TF_SessionRun(model->session, NULL,
		input, &input_value, 1,
		output, &output_value, 1,
		/* No target operations to run */
		NULL, 0, NULL, model->status);
	if (!Okay(model->status)) return NULL;
	TF_DeleteTensor(input_value);
	TF_DeleteTensor(output_value);
```

So when I run this code without `TF_SessionRun` I have no memory leaks. When I run this code with `TF_SessionRun` I get 3 new allocations that is not freed and has size described above. So when program continually predict it ran out of memory.

I have tried two models of different realizations of U-Net but all of them have same leak. The leak remains in both GPU and CPU dlls. Everything besides the leak works well and prediction gives result.

**Comment by StephenGreat** [comment]
@VladislavAD Did you solve this problem? I have similar issue too. My TF version is 1.12.

**Comment by VladislavAD** [comment]
@StephenGreat I solved it by using [this repo by Neargye](https://github.com/Neargye/hello_tf_c_api). I don't find what caused the leak, but my problem was solved. You can try to compare your code with provided in repo, I suppose the leak is caused by something in graph initialization.

**Comment by StephenGreat** [comment]
@VladislavAD Thanks for sharing the repo. I found my problems. That I have forgot to use TF_DeleteTensor() to delete output tensors. So the memory kept growing.

**Comment by adaber** [comment]
Hey everyone :)

I would just like to ask a question regarding TF_SessionRun and output tensors. It is directly related to the problem you encountered.

It seems to me that you have to free the previously used and allocate new output TF_Tensor object(s) every time you call TF_SessionRun. That seems very memory fragmentation heavy plus the operating system has to spend time allocating a new chunk of memory with every TF_SessionRun call. Is this true ?

I've recently started using the Tensorflow C library and may not be fully familiar with how it works.
 
Thanks :)

**Comment by StephenGreat** [comment]
@adaber In c_api.h, the comment about `TF_SessionRun()` shows something.
[c_api.zip](https://github.com/tensorflow/tensorflow/files/3523715/c_api.zip)
```
// On success, the tensors corresponding to outputs[0,noutputs-1] are placed in
// output_values[]. Ownership of the elements of output_values[] is transferred
// to the caller, **which must eventually call TF_DeleteTensor on them.**
```
In my experiment, I have to call TF_DeleteTensor to delete tenors after calling TF_SessionRun() to avoid the growing system memory.

**Comment by adaber** [comment]
Hey Stephen,

Thanks for the response.

I understand that part. You pass a pointer to a TF_Tensor pointer (output_values) which is used by TF_SessionRun to allocate/create TF_Tensor output objects. Then, you are responsible to delete/deallocate those objects.

My question was more about if we could reuse the same output TF_Tensor objects (output_values) in subsequent TF_SessionRun calls without having to deallocate and allocate new TF_Tensor objects every time we call TF_SessionRun ? We will, of course, delete them (once) in a destructor eventually.

I hope someone from the Tensorflow's team can also comment on this :)

Thanks.

**Comment by ZhuoranLyu** [comment]
> Hey Stephen,
> 
> Thanks for the response.
> 
> I understand that part. You pass a pointer to a TF_Tensor pointer (output_values) which is used by TF_SessionRun to allocate/create TF_Tensor output objects. Then, you are responsible to delete/deallocate those objects.
> 
> My question was more about if we could reuse the same output TF_Tensor objects (output_values) in subsequent TF_SessionRun calls without having to deallocate and allocate new TF_Tensor objects every time we call TF_SessionRun ? We will, of course, delete them (once) in a destructor eventually.
> 
> I hope someone from the Tensorflow's team can also comment on this :)
> 
> Thanks.

@adaber Same question, here is my code, wondering if this is the correct way to do so.

`for (int i = 0; i < 100; i++) { `
  `	  tf_utils::RunSession(session, input_ops, input_tensors, out_ops, output_tensors);  `
`	  tf_utils::DeleteTensors(output_tensors);  `
`}`

**Comment by adaber** [comment]
Hi ZhuoranLyu,

The order of operations seems ok. The overall correctness depends on the implementation of RunSession and DeleteTensors, of course. Those are probably two custom functions defined in a namespace.

**Comment by Weishuo93** [comment]
> Hey Stephen,
> 
> Thanks for the response.
> 
> I understand that part. You pass a pointer to a TF_Tensor pointer (output_values) which is used by TF_SessionRun to allocate/create TF_Tensor output objects. Then, you are responsible to delete/deallocate those objects.
> 
> My question was more about if we could reuse the same output TF_Tensor objects (output_values) in subsequent TF_SessionRun calls without having to deallocate and allocate new TF_Tensor objects every time we call TF_SessionRun ? We will, of course, delete them (once) in a destructor eventually.
> 
> I hope someone from the Tensorflow's team can also comment on this :)
> 
> Thanks.

@adaber 
Hi adaber, I came across the same problem with you. 
I tried in my program and print the address for "output_values". But it kept changing after each TF_SessionRun. 
So I am wondering if we can use a certain trunk of memory to hold the outputs. Have you figure out a solution?
Cheers!

**Comment by eirikaso** [comment]
@Weishuo93 
Hi. Did you find the solution? Having same issues here
